From c2849dce7ff45f6c003198b43153c7e2c597c988 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 22 Jul 2022 12:34:27 +1000
Subject: [PATCH] xactengine3_7: Remove Entry after FAudio Destroyed call

FACTWaveBank_Destroy will invoke the callback which we
attempt to lookup the wavebank.

The callback must have a pointer, help states it isn't valid but still
points to the wavebank that was destroyed.
---
 dlls/xactengine3_7/xact_dll.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/xactengine3_7/xact_dll.c b/dlls/xactengine3_7/xact_dll.c
index 19c02036907..a72afcc18e1 100644
--- a/dlls/xactengine3_7/xact_dll.c
+++ b/dlls/xactengine3_7/xact_dll.c
@@ -661,10 +661,11 @@ static HRESULT WINAPI IXACT3WaveBankImpl_Destroy(IXACT3WaveBank *iface)
 
     TRACE("(%p)\n", This);
 
+    hr = FACTWaveBank_Destroy(This->fact_wavebank);
+
     EnterCriticalSection(&This->engine->wb_wrapper_lookup_cs);
 
     entry = wine_rb_get(&This->engine->wb_wrapper_lookup, This->fact_wavebank);
-
     if (!entry)
     {
         LeaveCriticalSection(&This->engine->wb_wrapper_lookup_cs);
@@ -681,7 +682,6 @@ static HRESULT WINAPI IXACT3WaveBankImpl_Destroy(IXACT3WaveBank *iface)
         HeapFree(GetProcessHeap(), 0, lookup);
     }
 
-    hr = FACTWaveBank_Destroy(This->fact_wavebank);
     HeapFree(GetProcessHeap(), 0, This);
     return hr;
 }
-- 
2.35.1

